test: validate implementation against Julia test fixtures

anandkaranubc opened the pull request:
Resolves None

## Description

> What is the purpose of this pull request?

This pull request:

- Adds tests to validate `kernelCos` against the Julia implementation.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves None

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

anandkaranubc committed:
test: validate kernelCos against Julia implementation

anandkaranubc commented:
Once this is approved, I will do the same for `kernelSin` and `kernelTan`

anandkaranubc committed:
chore: update copyright years

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/kernel-cos/test/fixtures/julia/runner.jl:
```diff
@@ -0,0 +1,74 @@
+# Extract the directory in which this file resides:
+dir = dirname( file );
+
+# Values within the defined domain:
+x = range( -pi/4.0, stop = pi/4.0, length = 1000 )
+gen( x, "small_range.json" );
+
+# Positive values outside the defined domain:
+x = range( 40.0*pi/4.0, stop = 200*pi/4.0, length = 1000 )
+gen( x, "large_positive.json" );
+
+# Negative values outside the defined domain:
+x = range( -200*pi/4.0, stop = -40*pi/4.0, length = 1000 )
+gen( x, "large_negative.json" );
```
I hope the names of the files are fine.

kgryte replied:
Yes, this seems reasonable.

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/kernel-cos/test/test.native.js:
```diff
@@ -48,52 +54,64 @@ tape( 'main export is a function', opts, function test( t ) {
 		case 2:
 			out = -kernelCos( y[ 0 ], y[ 1 ] );
-			t.strictEqual( out, cos( x ), 'returns expected value' );
+			if ( out === expected[ i ] ) {
+				t.strictEqual( out, expected[ i ], 'returns expected value' );
+			} else {
+				delta = abs( out - expected[ i ] );
+				tol = EPS * abs( expected[ i ] );
+				t.ok( delta <= tol, 'within tolerance. x: '+x+'. out: '+out+'. E: '+expected[i]+'. tol: '+tol+'. Δ: '+delta+'.' );
+			}
```
There is only ONE value here that is very slightly different from the expected value (~1e-16). That's why I added this if-else block.

kgryte replied:
That is fine. Just add a comment mentioning the difference compared to JS.

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/kernel-cos/test/test.native.js:
```diff
@@ -103,26 +121,36 @@ tape( 'the function can be used to compute the cosine for input values outside o
 		case 2:
 			out = -kernelCos( y[ 0 ], y[ 1 ] );
-			t.strictEqual( out, cos( x ), 'returns expected value' );
+			if ( out === expected[ i ] ) {
+				t.strictEqual( out, expected[ i ], 'returns expected value' );
+			} else {
+				delta = abs( out - expected[ i ] );
+				tol = EPS * abs( expected[ i ] );
+				t.ok( delta <= tol, 'within tolerance. x: '+x+'. out: '+out+'. E: '+expected[i]+'. tol: '+tol+'. Δ: '+delta+'.' );
+			}
```
Same comment over here. Just one value.

kgryte replied:
Same comment.

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/kernel-cos/test/test.js:
```diff
@@ -72,7 +77,7 @@ tape( 'the function can be used to compute the cosine for input values outside o
 
-	values = linspace( 40.0*PI/4.0, 200*PI/4.0, 1000 );
+	values = largePositive.x;
 	y = new Array( 2 );
```
The linter is failing because of this. Is there a need to change this `Array` declaration?

kgryte replied:
@anandkaranubc Make `y = [ 0.0, 0.0 ]`.

anandkaranubc committed:
test: use correct expected values

anandkaranubc committed:
Merge branch 'stdlib-js:develop' into test/kernel-cos

anandkaranubc committed:
chore: apply suggestions from review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/kernel-cos/test/test.native.js:
```diff
@@ -48,52 +54,66 @@ tape( 'main export is a function', opts, function test( t ) {
+				delta = abs( out - expected[ i ] );
+
+				// NOTE: the tolerance here is larger than for the JavaScript implementation due to compiler optimizations which may be performed resulting in result divergence. For discussion, see https://github.com/stdlib-js/stdlib/pull/2298#discussion_r1624765205
+				tol = EPS * abs( expected[ i ] );
```
If `CFLAGS="-ffp-contract=off"` is used while building the native add-on, I can confirm that adding this is not needed.

kgryte reviewed (approved):
LGTM
